lib/src/install.rs

```diff
@@ -1423,6 +1525,502 @@ impl BoundImages {
+
+    // transport's display is already of type "<transport_type>:"
+    composefs_oci_pull(&Arc::new(repo), &format!("{transport}{image_name}"), None).await
+}
+
+fn get_booted_bls() -> Result<BLSConfig> {
+    let cmdline = crate::kernel::parse_cmdline()?;
+    let booted = cmdline.iter().find_map(|x| x.strip_prefix("composefs="));
+
+    let Some(booted) = booted else {
+        anyhow::bail!("Failed to find composefs parameter in kernel cmdline");
+    };
+
+    for entry in std::fs::read_dir("/sysroot/boot/loader/entries")? {
+        let entry = entry?;
+
+        if !entry.file_name().as_str()?.ends_with(".conf") {
+            continue;
+        }
+
+        let bls = parse_bls_config(&std::fs::read_to_string(&entry.path())?)?;
+
+        if bls.options.contains(booted) {
+            return Ok(bls);
+        }
+    }
+
+    Err(anyhow::anyhow!("Booted BLS not found"))
+}
+
+pub(crate) enum BootSetupType<'a> {
+    /// For initial setup, i.e. install to-disk
+    Setup(&'a RootSetup),
+    /// For `bootc upgrade`
+    Upgrade,
+}
+
+#[context("Setting up BLS boot")]
+pub(crate) fn setup_composefs_bls_boot(
+    setup_type: BootSetupType,
+    // TODO: Make this generic
+    repo: ComposefsRepository<Sha256HashValue>,
+    id: &Sha256HashValue,
+    entry: ComposefsBootEntry<Sha256HashValue>,
+) -> Result<()> {
+    let id_hex = id.to_hex();
+
+    let (root_path, cmdline_refs) = match setup_type {
+        BootSetupType::Setup(root_setup) => {
+            // root_setup.kargs has [root=UUID=<UUID>, "rw"]
+            let mut cmdline_options = String::from(root_setup.kargs.join(" "));
+            cmdline_options.push_str(&format!(" composefs={id_hex}"));
+
+            (root_setup.physical_root_path.clone(), cmdline_options)
+        }
+
+        BootSetupType::Upgrade => (
+            Utf8PathBuf::from("/sysroot"),
+            vec![
+                format!("root=UUID={DPS_UUID}"),
+                RW_KARG.to_string(),
+                format!("composefs={id_hex}"),
+            ]
+            .join(" "),
+        ),
+    };
+
+    let boot_dir = root_path.join("boot");
+
+    let bls_config = match &entry {
+        ComposefsBootEntry::Type1(..) => todo!(),
+        ComposefsBootEntry::Type2(..) => todo!(),
+        ComposefsBootEntry::UsrLibModulesUki(..) => todo!(),
+
+        ComposefsBootEntry::UsrLibModulesVmLinuz(usr_lib_modules_vmlinuz) => {
+            // Write the initrd and vmlinuz at /boot/<id>/
+            let path = boot_dir.join(&id_hex);
+            create_dir_all(&path)?;
+
+            let vmlinuz_path = path.join("vmlinuz");
+            let initrd_path = path.join("initrd");
+
+            std::fs::write(
+                &vmlinuz_path,
+                read_file(&usr_lib_modules_vmlinuz.vmlinuz, &repo).context("Reading vmlinuz")?,
+            )
+            .context("Writing vmlinuz to path")?;
+
+            if let Some(initramfs) = &usr_lib_modules_vmlinuz.initramfs {
+                std::fs::write(
```

Comment:
Again should use atomic writes - I think we also want to ensure we `fsync` here for sure.